File: src/validator.js
import fs from 'fs';
import path from 'path';

export class Validator {
  constructor(profilePathOrLang) {
    this.itemMap = new Map(); // Maps item ID to item data and source file
    this.levelMap = new Map(); // Maps level number to level data and source file
    this.profile = null;
    this.registry = null;

    if (profilePathOrLang) {
      if (typeof profilePathOrLang === 'string') {
        if (profilePathOrLang.endsWith('registry.json')) {
          this.registry = this.loadJSON(profilePathOrLang);
        } else {
          try {
            if (fs.existsSync(profilePathOrLang)) {
              const loaded = this.loadJSON(profilePathOrLang);
              if (loaded.languages) {
                this.registry = loaded;
              } else {
                this.profile = loaded;
              }
            } else {
              const registryPath = '/app/profiles/registry.json';
              this.registry = this.loadJSON(registryPath);
              this.profile = this.getProfileForLanguage(profilePathOrLang);
            }
          } catch (e) {
            try {
              const registryPath = '/app/profiles/registry.json';
              this.registry = this.loadJSON(registryPath);
              this.profile = this.getProfileForLanguage(profilePathOrLang);
            } catch (innerErr) {
              throw new Error(`Failed to initialize Validator: ${profilePathOrLang}. Error: ${e.message}`);
            }
          }
        }
      } else if (typeof profilePathOrLang === 'object') {
        this.profile = profilePathOrLang;
      }
    } else {
      const registryPath = '/app/profiles/registry.json';
      this.registry = this.loadJSON(registryPath);
    }
  }

  loadJSON(filePath) {
    try {
      const data = fs.readFileSync(filePath, 'utf8');
      return JSON.parse(data);
    } catch (e) {
      throw new Error(`Failed to load or parse JSON from ${filePath}: ${e.message}`);
    }
  }

  getProfileForLanguage(langCode) {
    if (!this.registry) {
      const registryPath = '/app/profiles/registry.json';
      this.registry = this.loadJSON(registryPath);
    }
    const cleanCode = langCode.toLowerCase();
    let entry = this.registry.languages[cleanCode];
    if (!entry) {
      entry = Object.values(this.registry.languages).find(l => l.language.toLowerCase() === cleanCode);
    }
    if (!entry) {
      throw new Error(`Language '${langCode}' is not explicitly registered in the unified registry.`);
    }

    const fallbackProfile = {
      language: entry.language || cleanCode,
      iso_code: entry.iso_code || cleanCode,
      permitted_genders: entry.permitted_genders || [],
      active_phonetic_systems: entry.active_phonetic_systems || [],
      script_traits: {
        is_abugida: entry.script_traits?.is_abugida ?? false,
        has_matra: entry.script_traits?.has_matra ?? false,
        has_sound_family: entry.script_traits?.has_sound_family ?? false,
        has_tones: entry.script_traits?.has_tones ?? false,
        has_word_masks: entry.script_traits?.has_word_masks ?? false,
        ...(entry.script_traits || {})
      },
      script_boundaries: entry.script_boundaries || { ranges: [] }
    };
    return fallbackProfile;
  }

  detectLanguage(data) {
    if (!this.registry) {
      const registryPath = '/app/profiles/registry.json';
      this.registry = this.loadJSON(registryPath);
    }

    const strData = JSON.stringify(data);
    for (const [langCode, langObj] of Object.entries(this.registry.languages)) {
      if (data[`module_name_${langCode}`] !== undefined || strData.includes(`title_${langCode}`)) {
        return langCode;
      }
    }

    const charCounts = {};
    for (const langCode of Object.keys(this.registry.languages)) {
      charCounts[langCode] = 0;
    }

    const scanStrings = (val) => {
      if (typeof val === 'string') {
        for (const [langCode, langObj] of Object.entries(this.registry.languages)) {
          const ranges = langObj.script_boundaries?.ranges || [];
          for (const [start, end] of ranges) {
            const startCode = start.codePointAt(0);
            const endCode = end.codePointAt(0);
            for (let i = 0; i < val.length; i++) {
              const code = val.codePointAt(i);
              if (code >= startCode && code <= endCode) {
                const char = String.fromCodePoint(code);
                if (startCode > 0x7F || /^\p{L}$/u.test(char)) {
                  charCounts[langCode]++;
                }
              }
            }
          }
        }
      } else if (Array.isArray(val)) {
        for (const item of val) {
          scanStrings(item);
        }
      } else if (val && typeof val === 'object') {
        for (const value of Object.values(val)) {
          scanStrings(value);
        }
      }
    };

    scanStrings(data);

    const nonLatinLanguages = {};
    const latinLanguages = {};
    for (const [langCode, langObj] of Object.entries(this.registry.languages)) {
      const ranges = langObj.script_boundaries?.ranges || [];
      const hasNonLatinRange = ranges.some(([start, end]) => start.codePointAt(0) > 0x2FF);
      const count = charCounts[langCode] || 0;
      if (hasNonLatinRange) {
        nonLatinLanguages[langCode] = count;
      } else {
        latinLanguages[langCode] = count;
      }
    }

    const hasNonLatinMatch = Object.values(nonLatinLanguages).some(count => count > 0);
    const activeCandidates = hasNonLatinMatch ? nonLatinLanguages : latinLanguages;

    let bestLang = null;
    let maxCount = 0;
    for (const [langCode, count] of Object.entries(activeCandidates)) {
      if (count > maxCount) {
        maxCount = count;
        bestLang = langCode;
      }
    }

    return bestLang;
  }

  validateScriptBoundaries(text, langCode) {
    const profile = this.getProfileForLanguage(langCode);
    const ranges = profile.script_boundaries?.ranges || [];
    if (ranges.length === 0) return;

    for (let i = 0; i < text.length; i++) {
      const code = text.codePointAt(i);
      let allowed = false;
      for (const [start, end] of ranges) {
        const startCode = start.codePointAt(0);
        const endCode = end.codePointAt(0);
        if (code >= startCode && code <= endCode) {
          allowed = true;
          break;
        }
      }

      if (!allowed) {
        const char = String.fromPoint ? String.fromPoint(code) : String.fromCodePoint(code);
        const isLetter = /^\p{L}$/u.test(char);
        if (isLetter) {
          throw new Error(`Script boundary violation: character '${char}' (U+${code.toString(16).toUpperCase()}) in target '${text}' is outside the permitted script boundaries for language '${langCode}'.`);
        }
      }
    }
  }

  // Pre-loads levels so referential integrity can be verified across all files in the module
  registerLevels(levelsFilePath) {
    const data = this.loadJSON(levelsFilePath);
    const levels = data.levels || [];
    const errors = [];

    // Basic levels array check
    if (!Array.isArray(levels)) {
      errors.push({
        file: levelsFilePath,
        message: "The 'levels' field must be an array."
      });
      return errors;
    }

    for (const lvl of levels) {
      const lvlNum = lvl.level;
      if (typeof lvlNum !== 'number') {
        errors.push({
          file: levelsFilePath,
          message: `Level number must be a number, got: ${typeof lvlNum}`
        });
        continue;
      }

      if (this.levelMap.has(lvlNum)) {
        errors.push({
          file: levelsFilePath,
          message: `Duplicate level number found: ${lvlNum}`
        });
      }
      this.levelMap.set(lvlNum, { data: lvl, file: levelsFilePath });

      const items = lvl.items || [];
      if (!Array.isArray(items)) {
        errors.push({
          file: levelsFilePath,
          level: lvlNum,
          message: `The 'items' field in level ${lvlNum} must be an array.`
        });
        continue;
      }

      // Rigid Lesson Counting Limit: exactly three items per level
      if (items.length !== 3) {
        errors.push({
          file: levelsFilePath,
          level: lvlNum,
          message: `Rigid lesson count limit violated in level ${lvlNum}: expected exactly 3 items, found ${items.length}`
        });
      }

      // Check for summary recap
      if (!lvl.summary) {
        errors.push({
          file: levelsFilePath,
          level: lvlNum,
          message: `Level ${lvlNum} is missing a summary recap.`
        });
      } else {
        const sum = lvl.summary;
        if (typeof sum.recap_en !== 'string' || !Array.isArray(sum.key_forms) || typeof sum.next_up_en !== 'string') {
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            message: `Level ${lvlNum} summary recap is invalid: must contain 'recap_en' (string), 'key_forms' (array of strings), and 'next_up_en' (string).`
          });
        }
      }

      for (const item of items) {
        if (!item || typeof item.id !== 'string') {
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            message: `Found an item without a valid string ID in level ${lvlNum}`
          });
          continue;
        }

        if (this.itemMap.has(item.id)) {
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            itemId: item.id,
            message: `Duplicate item ID found: ${item.id}`
          });
        }
        this.itemMap.set(item.id, { data: item, level: lvlNum, file: levelsFilePath });
      }
    }

    return errors;
  }

  validateLevelFile(levelsFilePath) {
    const data = this.loadJSON(levelsFilePath);
    const levels = data.levels || [];
    const errors = [];

    let profile = this.profile;
    let langCode = profile ? (profile.iso_code || profile.language) : null;
    if (!langCode) {
      langCode = this.detectLanguage(data);
    }
    if (!langCode) {
      throw new Error(`Language could not be detected, and none is configured. Failing closed.`);
    }
    profile = this.getProfileForLanguage(langCode);

    for (const lvl of levels) {
      const lvlNum = lvl.level;
      const items = lvl.items || [];

      for (const item of items) {
        if (!item || typeof item.id !== 'string') continue;

        if (item.target) {
          try {
            this.validateScriptBoundaries(item.target, langCode);
          } catch (err) {
            errors.push({
              file: levelsFilePath,
              level: lvlNum,
              itemId: item.id,
              message: err.message
            });
          }
        }

        // Polymorphic schema validations based on Language Profile
        const genders = profile.permitted_genders || [];
        if (item.gender !== null && item.gender !== undefined) {
          if (!genders.includes(item.gender)) {
            errors.push({
              file: levelsFilePath,
              level: lvlNum,
              itemId: item.id,
              message: `Gender '${item.gender}' is not permitted in language '${profile.language}'. Allowed genders: [${genders.join(', ')}]`
            });
          }
        } else if (item.type === 'word' && genders.length > 0 && profile.language === 'hindi') {
          // If the profile mandates noun genders (e.g. Hindi), warn/error if missing
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            itemId: item.id,
            message: `Missing required gender field for Hindi word: '${item.target}'`
          });
        }

        // Script traits checks
        const isAbugida = profile.script_traits?.is_abugida;
        if (item.type === 'letter') {
          if (isAbugida) {
            // For Abugida scripts, fields like 'matra' and 'sound_family' must be defined (can be null but key must exist)
            if (!('matra' in item)) {
              errors.push({
                file: levelsFilePath,
                level: lvlNum,
                itemId: item.id,
                message: `Abugida script requires 'matra' field on letter items, but it was omitted.`
              });
            }
            if (!('sound_family' in item)) {
              errors.push({
                file: levelsFilePath,
                level: lvlNum,
                itemId: item.id,
                message: `Abugida script requires 'sound_family' field on letter items, but it was omitted.`
              });
            }
          }
          // Note: for non-Abugida scripts, the absence of 'matra' or 'sound_family' is perfectly valid and ignored.
        }

        // Validate basic fields of learning/practice/game cards
        if (!item.cards) {
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            itemId: item.id,
            message: `Item ${item.id} is missing 'cards' block.`
          });
          continue;
        }

        const cards = item.cards;
        if (!cards.learn || !cards.practice || !cards.game) {
          errors.push({
            file: levelsFilePath,
            level: lvlNum,
            itemId: item.id,
            message: `Item ${item.id} cards must contain 'learn', 'practice', and 'game' blocks.`
          });
        }

        // Categorization game validations (e.g., sorting game using categorization metadata)
        if (cards.game && (cards.game.format === 'sorting' || cards.game.engine === 'sorting')) {
          const game = cards.game;
          if (!game.categories || !Array.isArray(game.categories)) {
            errors.push({
              file: levelsFilePath,
              level: lvlNum,
              itemId: item.id,
              message: `Sorting game configuration in item ${item.id} is missing 'categories' array.`
            });
          }
          if (!game.items || !Array.isArray(game.items)) {
            errors.push({
              file: levelsFilePath,
              level: lvlNum,
              itemId: item.id,
              message: `Sorting game configuration in item ${item.id} is missing sorted 'items' array.`
            });
          } else {
            for (const gItem of game.items) {
              if (typeof gItem.text !== 'string' || typeof gItem.category !== 'string') {
                errors.push({
                  file: levelsFilePath,
                  level: lvlNum,
                  itemId: item.id,
                  message: `Sorting game item must contain 'text' and 'category' strings.`
                });
              } else if (game.categories && !game.categories.includes(gItem.category)) {
                errors.push({
                  file: levelsFilePath,
                  level: lvlNum,
                  itemId: item.id,
                  message: `Sorting game item category '${gItem.category}' is not listed in 'categories' [${game.categories.join(', ')}]`
                });
              }
            }
          }
        }
      }
    }

    return errors;
  }

  validateCheckpointFile(checkpointFilePath) {
    const data = this.loadJSON(checkpointFilePath);
    const checkpoints = data.checkpoints || [];
    const errors = [];

    for (const cp of checkpoints) {
      const cpNum = cp.checkpoint;
      const questions = cp.questions || [];

      if (!Array.isArray(questions)) {
        errors.push({
          file: checkpointFilePath,
          checkpoint: cpNum,
          message: `The 'questions' field in checkpoint ${cpNum} must be an array.`
        });
        continue;
      }

      for (const q of questions) {
        if (!q || typeof q.id !== 'string') {
          errors.push({
            file: checkpointFilePath,
            checkpoint: cpNum,
            message: `Checkpoint ${cpNum} has a question missing a valid string ID.`
          });
          continue;
        }

        // Referential Integrity check: All checkpoint questions must map back to source items
        const sourceIds = q.source_item_ids || [];
        if (!Array.isArray(sourceIds) || sourceIds.length === 0) {
          errors.push({
            file: checkpointFilePath,
            checkpoint: cpNum,
            questionId: q.id,
            message: `Question ${q.id} in checkpoint ${cpNum} must have non-empty 'source_item_ids' array.`
          });
        } else {
          for (const sId of sourceIds) {
            if (!this.itemMap.has(sId)) {
              errors.push({
                file: checkpointFilePath,
                checkpoint: cpNum,
                questionId: q.id,
                message: `Referential Integrity violation: Question ${q.id} references non-existent source level item '${sId}'`
              });
            }
          }
        }
      }
    }

    return errors;
  }

  validateMockFile(mockFilePath) {
    const data = this.loadJSON(mockFilePath);
    const errors = [];
    const mockExam = data.mock_exam || data; // Mock exams can have nested or direct mock_exam block
    const questions = mockExam.questions || [];

    if (!Array.isArray(questions)) {
      errors.push({
        file: mockFilePath,
        message: "Mock exam must have an array of 'questions'."
      });
      return errors;
    }

    for (const q of questions) {
      if (!q || typeof q.id !== 'string') {
        errors.push({
          file: mockFilePath,
          message: "Mock exam has a question missing a valid string ID."
        });
        continue;
      }

      // Referential Integrity check: All mock exam questions must map back to source items
      const sourceIds = q.source_item_ids || [];
      if (!Array.isArray(sourceIds) || sourceIds.length === 0) {
        errors.push({
          file: mockFilePath,
          questionId: q.id,
          message: `Question ${q.id} in mock exam must have non-empty 'source_item_ids' array.`
        });
      } else {
        for (const sId of sourceIds) {
          if (!this.itemMap.has(sId)) {
            errors.push({
              file: mockFilePath,
              questionId: q.id,
              message: `Referential Integrity violation: Mock exam question ${q.id} references non-existent source level item '${sId}'`
            });
          }
        }
      }
    }

    return errors;
  }

  validateSpineFile(spineFilePath) {
    const data = this.loadJSON(spineFilePath);
    const errors = [];
    const spineItems = data.spine || [];

    if (!Array.isArray(spineItems)) {
      errors.push({
        file: spineFilePath,
        message: "Spine file must contain an array of spine items under 'spine'."
      });
      return errors;
    }

    let profile = this.profile;
    let langCode = profile ? (profile.iso_code || profile.language) : null;
    if (!langCode) {
      langCode = this.detectLanguage(data);
    }

    for (const sItem of spineItems) {
      if (!sItem || typeof sItem.id !== 'string') {
        errors.push({
          file: spineFilePath,
          message: "Spine entry is missing a valid string ID."
        });
        continue;
      }

      if (sItem.target && langCode) {
        try {
          this.validateScriptBoundaries(sItem.target, langCode);
        } catch (err) {
          errors.push({
            file: spineFilePath,
            itemId: sItem.id,
            message: err.message
          });
        }
      }

      // Referential Integrity with level files
      if (!this.itemMap.has(sItem.id)) {
        // If it's a partial spine (like module 9 with levels missing), check if the level is marked as missing
        const isMissingLevel = data.levels_missing?.includes(sItem.level);
        if (!isMissingLevel) {
          errors.push({
            file: spineFilePath,
            itemId: sItem.id,
            message: `Referential Integrity violation: Spine item '${sItem.id}' references level ${sItem.level} but does not exist in any registered level files.`
          });
        }
        continue;
      }

      const registeredItem = this.itemMap.get(sItem.id).data;
      const registeredLvl = this.itemMap.get(sItem.id).level;

      // Validate corresponding attributes
      if (sItem.level !== registeredLvl) {
        errors.push({
          file: spineFilePath,
          itemId: sItem.id,
          message: `Spine item '${sItem.id}' level (${sItem.level}) does not match levels file level (${registeredLvl})`
        });
      }

      if (sItem.type && sItem.type !== registeredItem.type) {
        errors.push({
          file: spineFilePath,
          itemId: sItem.id,
          message: `Spine item '${sItem.id}' type (${sItem.type}) does not match levels file type (${registeredItem.type})`
        });
      }

      if (sItem.target && sItem.target !== registeredItem.target) {
        errors.push({
          file: spineFilePath,
          itemId: sItem.id,
          message: `Spine item '${sItem.id}' target (${sItem.target}) does not match levels file target (${registeredItem.target})`
        });
      }

      if (sItem.gender !== undefined && sItem.gender !== registeredItem.gender) {
        errors.push({
          file: spineFilePath,
          itemId: sItem.id,
          message: `Spine item '${sItem.id}' gender (${sItem.gender}) does not match levels file gender (${registeredItem.gender})`
        });
      }
    }

    return errors;
  }
}

File: core_engine/validator.py
import os
from core_engine.sandbox import exec_sandboxed
from plugins.base import BasePlugin
from core_engine.schema import validate_structure

class Validator:
    """
    Core validator coordinating schema validation and target-language pluggable rules.
    """
    def __init__(self, plugins_dir="/app/plugins"):
        self.plugins_dir = plugins_dir
        self._plugin_cache = {}

    def load_plugin(self, language: str) -> BasePlugin:
        """
        Dynamically loads the plugin for the given language inside the restricted sandbox.
        """
        if not language:
            return None
            
        language = language.lower()
        if language in self._plugin_cache:
            return self._plugin_cache[language]

        plugin_file = os.path.join(self.plugins_dir, f"{language}_plugin.py")
        if not os.path.exists(plugin_file):
            return None

        with open(plugin_file, "r", encoding="utf-8") as f:
            source = f.read()

        # Execute inside the safe sandboxed environment
        extra_globals = {
            "BasePlugin": BasePlugin
        }
        sandbox_globals = exec_sandboxed(source, plugin_file, extra_globals)

        # Retrieve the dynamic subclass of BasePlugin
        plugin_class = None
        for name, obj in sandbox_globals.items():
            if isinstance(obj, type) and issubclass(obj, BasePlugin) and obj is not BasePlugin:
                plugin_class = obj
                break

        if not plugin_class:
            raise RuntimeError(f"No valid class inheriting from BasePlugin found in {plugin_file}")

        plugin_instance = plugin_class()
        self._plugin_cache[language] = plugin_instance
        return plugin_instance

    def _load_registry(self):
        import json
        registry_path = "/app/profiles/registry.json"
        if not os.path.exists(registry_path):
            raise RuntimeError(f"Unified registry file not found at {registry_path}")
        with open(registry_path, "r", encoding="utf-8") as f:
            self.registry = json.load(f)

    def get_profile(self, language: str) -> dict:
        if not hasattr(self, "registry") or not self.registry:
            self._load_registry()
            
        language = language.lower()
        if language not in self.registry.get("languages", {}):
            # Try finding by full name
            for k, v in self.registry.get("languages", {}).items():
                if v.get("language", "").lower() == language:
                    language = k
                    break
                    
        if language not in self.registry.get("languages", {}):
            raise ValueError(f"Language '{language}' is not explicitly registered in the unified registry. Failing closed.")
            
        profile = self.registry["languages"][language]
        
        # Structured Fallback Profile: provide default values for any missing sub-rules
        fallback_profile = {
            "language": profile.get("language", language),
            "iso_code": profile.get("iso_code", language),
            "permitted_genders": profile.get("permitted_genders", []),
            "active_phonetic_systems": profile.get("active_phonetic_systems", []),
            "script_traits": {
                "is_abugida": False,
                "has_matra": False,
                "has_sound_family": False,
                "has_tones": False,
                "has_word_masks": False
            },
            "script_boundaries": {
                "ranges": []
            }
        }
        if "script_traits" in profile:
            fallback_profile["script_traits"].update(profile["script_traits"])
        if "script_boundaries" in profile:
            fallback_profile["script_boundaries"].update(profile["script_boundaries"])
            
        return fallback_profile

    def detect_language(self, data: dict) -> str:
        """
        Detects language from content or metadata patterns using character block analysis.
        """
        if not hasattr(self, "registry") or not self.registry:
            self._load_registry()
            
        # Check explicit language keys
        for lang_code, lang_obj in self.registry.get("languages", {}).items():
            if f"module_name_{lang_code}" in data or f"title_{lang_code}" in str(data):
                return lang_code
                
        # Character block analysis over text values
        char_counts = {lang_code: 0 for lang_code in self.registry.get("languages", {})}
        
        def count_chars(v):
            if isinstance(v, str):
                for lang_code, lang_obj in self.registry.get("languages", {}).items():
                    ranges = lang_obj.get("script_boundaries", {}).get("ranges", [])
                    for start, end in ranges:
                        for char in v:
                            if start <= char <= end:
                                if start > "\u007F" or char.isalpha():
                                    char_counts[lang_code] += 1
            elif isinstance(v, list):
                for item in v:
                    count_chars(item)
            elif isinstance(v, dict):
                for item in v.values():
                    count_chars(item)
                    
        count_chars(data)
        
        # Priority rule: if any non-Latin language (starts above \u02FF) has a match > 0, we only consider those.
        non_latin_languages = {}
        latin_languages = {}
        for lang_code, count in char_counts.items():
            lang_obj = self.registry["languages"][lang_code]
            ranges = lang_obj.get("script_boundaries", {}).get("ranges", [])
            has_non_latin_range = any(start > "\u02FF" for start, end in ranges)
            if has_non_latin_range:
                non_latin_languages[lang_code] = count
            else:
                latin_languages[lang_code] = count
                
        if any(count > 0 for count in non_latin_languages.values()):
            active_candidates = {k: v for k, v in non_latin_languages.items() if v > 0}
        else:
            active_candidates = latin_languages
            
        best_lang = None
        max_count = 0
        for lang_code, count in active_candidates.items():
            if count > max_count:
                max_count = count
                best_lang = lang_code
                
        return best_lang

    def validate_script_boundaries(self, data: dict, language: str) -> None:
        profile = self.get_profile(language)
        ranges = profile.get("script_boundaries", {}).get("ranges", [])
        if not ranges:
            return
            
        def check_item(item):
            target = item.get("target")
            if target and isinstance(target, str):
                for char in target:
                    # Check if char is in allowed ranges
                    allowed = False
                    for start, end in ranges:
                        if start <= char <= end:
                            allowed = True
                            break
                    if not allowed:
                        # Allow safe characters (space, digits, punctuation)
                        # and ONLY raise error if the character is an alphabetic letter
                        if char.isalpha():
                            raise ValueError(
                                f"Script boundary violation: character '{char}' (U+{ord(char):04X}) "
                                f"in target '{target}' is outside the permitted script boundaries for language '{language}'."
                            )
                            
        def traverse(v):
            if isinstance(v, dict):
                if "target" in v:
                    check_item(v)
                for val in v.values():
                    traverse(val)
            elif isinstance(v, list):
                for item in v:
                    traverse(item)
                    
        traverse(data)

    def validate(self, data: dict, schema_type: str, language: str = None) -> dict:
        """
        Executes normalization, core schema validation, and dynamic plugin rules.
        """
        if not hasattr(self, "registry") or not self.registry:
            self._load_registry()
            
        if language is None:
            language = self.detect_language(data)
            
        if not language:
            raise ValueError("Language could not be detected, and none was specified. Failing closed.")
            
        # Normalize language to registry key
        language = language.lower()
        profile = self.get_profile(language)
        lang_code = profile["iso_code"]
        
        # Enforce fail-closed structure: if no valid profile exists, or if a validation plugin is expected but missing, fail closed.
        plugin = None
        plugin_file = os.path.join(self.plugins_dir, f"{lang_code}_plugin.py")
        if not os.path.exists(plugin_file):
            plugin_file = os.path.join(self.plugins_dir, f"{profile['language']}_plugin.py")
            
        if os.path.exists(plugin_file):
            plugin = self.load_plugin(profile["language"] if f"{profile['language']}_plugin.py" in plugin_file else lang_code)
            
        # Preprocess/Normalize legacy structures
        if plugin:
            normalized_data = plugin.normalize(data)
        else:
            normalized_data = data
            
        # Pure structural validation using generic schemas (replaces static database/table check)
        validate_structure(normalized_data, schema_type)
        
        # Dynamic script boundary checks
        self.validate_script_boundaries(normalized_data, lang_code)
        
        # Dynamic semantic validation via sandboxed plugin
        if plugin:
            plugin.validate(normalized_data, schema_type)
            
        return normalized_data
